src/lib/storage/dictionary_segment.cpp

```diff
@@ -122,6 +123,207 @@ const ValueID DictionarySegment<T>::null_value_id() const {
   return _null_value_id;
 }
 
+template <typename T>
+ChunkOffset DictionarySegment<T>::get_non_null_begin_offset(
+    const std::shared_ptr<const PosList>& position_filter) const {
```

Comment:
You have added `advance` to all the iterators, right? Why do DictionarySegment, ValueSegment, and potentially all other Segment Types need their own implementation of get_non_null_begin_offset, get_non_null_end_offset, get_first_offset and get_last_offset? Would it be possible to write a set of template functions doing this? I think we have a strong interest in keeping the Segment classes as lean as possible, as they are such fundamental core structures.

Reply:
@cmfcmf and team, is it understandable what I meant by this? I think it's a rather important point

Reply:
Yes. We've been thinking about this and you are probably right. We'll play around with it in the following days/next week.

Reply:
We have implemented this in 14a3692f7ef3d2e9f4b292a7ecd974c982acab83, but are still facing the issue described in #1512.